src/ess/imaging/normalize.py

```diff
@@ -0,0 +1,235 @@
+
+    .. math::
+
+        Background = mean(OpenBeam, 'time') - mean(DarkCurrent, 'time')
+
+
+    Params
+    ------
+    oepn_beam:
+        Open beam image stack.
+
+    dark_current:
+        Dark current image.
+
+    background_threshold:
+        Threshold for the background pixel values.
+        Any pixel values less than ``background_threshold``
+        are replaced with ``background_threshold``.
+        Default is 1.0[counts].
+
+
+    """
+    diff = open_beam - dark_current
+    diff.data = sc.where(
+        diff.data < background_threshold, background_threshold, diff.data
+    )
+    return BackgroundImage(diff)
+
+
+SamplePixelThreshold = NewType("SamplePixelThreshold", sc.Variable)
+"""Threshold for the sample pixel values."""
+
+DEFAULT_SAMPLE_THRESHOLD = SamplePixelThreshold(sc.scalar(0.0, unit="counts"))
+
+
+def cleanse_sample_images(
+    sample_images: SampleImageStacks,
+    dark_current: DarkCurrentImage,
+    sample_threshold: SamplePixelThreshold = DEFAULT_SAMPLE_THRESHOLD,
+) -> CleansedSampleImages:
+    """Cleanse the sample image stack.
+
+    We subtract the averaged dark current image from the sample image stack.
+
+    .. math::
+
+        CleansedSample_{i} = Sample_{i} - mean(DarkCurrent, dim='time')
+
+        \\text{where } i \\text{ is an index of an image.}
+
+    Params
+    ------
+    sample_images:
+        Sample image stack.
+    dark_current:
+        Dark current image.
+    sample_threshold:
+        Threshold for the sample pixel values.
+        Any pixel values less than ``sample_threshold``
+        are replaced with ``sample_threshold``.
+        Default is 0.0[counts].
+
+    """
+    diff = sample_images - dark_current
+    diff.data = sc.where(diff.data < sample_threshold, sample_threshold, diff.data)
+    return CleansedSampleImages(diff)
+
+
+def average_background_pixel_counts(
+    background: BackgroundImage,
+) -> AverageBackgroundPixelCounts:
+    """Calculate the average background pixel counts."""
+    return AverageBackgroundPixelCounts(background.data.mean())
+
+
+def average_sample_pixel_counts(
+    sample_images: CleansedSampleImages, dark_current: DarkCurrentImage
+) -> AverageSamplePixelCounts:
+    """Calculate the average sample pixel counts.
+
+    Note that we calculate the mean of sample images and dark current images
+    first and subtract them afterwards,
+    instead of using the subtracted image stack directly.
+    This is for performance reasons, as the integer operation is faster than
+    the floating point operation.
+
+    Also, we don't calculate ``mean(sample_images)`` at once
+    since it is a large array and it may cause memory issues.
+
+    There was an example of 361 images of 2048x2048 pixels with 32-bit integer data
+    exceeded the limit of the maximum integer so the average calculation failed
+    and returned negative values.
+    """
+    return AverageSamplePixelCounts(
+        _mean_all_dims(sample_images.data) - dark_current.data.mean()
```

Comment:
I am wondering if sums should be used instead of means for both the `AverageBackgroundPixelCounts` and the `AverageSamplePixelCounts`?

Usually when we normalize by e.g. a monitor, we sum the counts?

Reply:
But we don't take same number of dark-current/open-beam images as the sample images.

Reply:
Probably best to discuss in person / over a call, as I am wondering if the reduction over the time dimension should also be sums instead of means?

If I understand correctly, it's to make sure everything has been scaled to the same integration time, before performing the division, right?

Reply:
> If I understand correctly, it's to make sure everything has been scaled to the same integration time, before performing the division, right?

Yes, that was the idea.

Reply:
Maybe it should be more like:

`(sample_image / sample_image_integration_time - dark_frame / dark_frame_integration_time) / (openbeam_image / openbeam_image_integration_time - dark_frame / dark_frame_integration_time)`

?

Reply:
Number of images is same as the `integration_time` since the shutter(?) speed is fixed.

Reply:
Ah so you're saying that doing a mean is essentially doing `image.sum(time) / integration_time` since you divide by the number of images (there is just a factor of `time_shutter_is_open` which cancels out everywhere)?

I guess if the shutter speed is adjustable this should be verified?

Are the dark frames and open beam always taken just before the sample is, or is a dark frame sometimes taken weeks in advanced and used for several experiments in a row?

If it's the former, then I guess we are fine with means.

Reply:
> Ah so you're saying that doing a mean is essentially doing image.sum(time) / integration_time since you divide by the number of images (there is just a factor of time_shutter_is_open which cancels out everywhere)?

Yes...! That's what I meant.
But maybe I should check with Robin once more and mention this in the docstring.

> I guess if the shutter speed is adjustable this should be verified?

Yes. Totally.

> Are the dark frames and open beam always taken just before the sample is, or is a dark frame sometimes taken weeks in advanced and used for several experiments in a row?

They will be taken again every time just before or after the sample images.

Background images(dark-current and open-beam) takes much less time than samples and
Odin team wanted every types of images in one file.

> If it's the former, then I guess we are fine with means.

Yes. In either case, I'll update the docstring.

Reply:
I double checked with Søren and opened an issue as a reminder.

https://github.com/scipp/essimaging/issues/48

And it gives warning in the average/normalization functions about exposure time.

I chose `warning` over `docstring` because I think we should use `exposure-time` as you said,
and it should be avilable in the file.